### ISSUE TITLE
Unsub unnecessarily triggers disconnect

### ISSUE THREAD

**Issue by richie3366** [adonisjs/adonis-redis#18]
## Package version
`2.0.7`

## Node.js and npm version
`v10.12.0` (current LTS)

## Sample Code
```javascript
'use strict'

const { Command } = require('@adonisjs/ace')

const Redis = use('Redis')
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms))

class TestRedisIssue extends Command {
  static get signature () {
    return 'test:redis:issue'
  }

  static get description () {
    return 'Tell something helpful about this command'
  }

  async handle (args, options) {
    this.info('Dummy implementation for test:redis:issue command')

    // Note: the adonis 4.1 docs (wrongly) says "message, channel, pattern"
    Redis.psubscribe('my-pattern-event:*', (pattern, message, channel) => { 
      console.log('redis:pmessage', pattern, message, channel)
    })

    Redis.subscribe('some-temp-channel', message => {
      console.log('redis:message:some-temp-channel', message)
    })

    await sleep(100)
    Redis.publish('my-pattern-event:test', 'test') // received
    await sleep(100)
    Redis.publish('some-temp-channel', 'some-message') // received
    await sleep(100)
    Redis.unsubscribe('some-temp-channel') // triggers the issue
    await sleep(100)
    Redis.publish('my-pattern-event:test2', 'test2') // not received! (this is bad since we did not unsubscribed from this pattern)
    await sleep(100)
    Redis.subscribe('some-other-queue', () => {}) // error:
    /*
      2018-10-26T12:01:37.961Z - warning:
        WARNING: Adonis has detected an unhandled promise rejection, which may
        cause undesired behavior in production.
        To stop this warning, use catch() on promises or wrap await
        calls inside try/catch.

      Error: Connection is closed.
          at Redis.sendCommand (/home/richie/dev/(redacted)/node_modules/ioredis/lib/redis.js:558:20)
          at Redis.subscribe (/home/richie/dev/(redacted)/node_modules/ioredis/lib/commander.js:131:17)
      From previous event:
          at Command.initPromise (/home/richie/dev/(redacted)/node_modules/ioredis/lib/command.js:63:18)
          at new Command (/home/richie/dev/(redacted)/node_modules/ioredis/lib/command.js:51:8)
          at Redis.subscribe (/home/richie/dev/(redacted)/node_modules/ioredis/lib/commander.js:131:29)
          at Promise (/home/richie/dev/(redacted)/node_modules/@adonisjs/redis/src/RedisFactory/index.js:218:33)
          at new Promise (<anonymous>)
          at Proxy.subscribe (/home/richie/dev/(redacted)/node_modules/@adonisjs/redis/src/RedisFactory/index.js:196:12)
          at TestRedisIssue.handle (/home/richie/dev/(redacted)/app/Commands/TestRedisIssue.js:37:11)
    */
  }
}

module.exports = TestRedisIssue
```

## Heart of the issue
Here:
https://github.com/adonisjs/adonis-redis/blob/03b45cfd379130fce981c2ab16b7f6a0504e7185/src/RedisFactory/index.js#L303

It checks if there is any other registered channel sub, but it doesn't check for any pchannel (different variable), so it disconnects the redis-sub connection.

And the reason of the triggered error is: the disconnect process doesn't clear the `this.subscriberConnection` variable ; auto-connect is consequently not called :
https://github.com/adonisjs/adonis-redis/blob/03b45cfd379130fce981c2ab16b7f6a0504e7185/src/RedisFactory/index.js#L165

My advice: do not close the connection, redis is capable of handling many many ones (https://redis.io/topics/clients#maximum-number-of-clients), a double connection from each instance would not matter that much (plus it was a needed slot just before closing it). If an adonis instance is based on a sub/unsub workflow, it will uselessly connects and disconnects to the redis server each time there is 0 listened channels, not a stable behavior.

**Comment by thetutlage** [comment]
**it will uselessly connects and disconnects to the redis server each time there is 0 listened channels, not a stable behavior.**

I understand the connecting and disconnecting can be avoided here, but why it's not a stable behavior?

**Comment by richie3366** [comment]
I mean, if I want to have a worker (through an adonis Command) relying on temporary channel subscriptions, my worker will constantly connect & disconnect whenever he needs to subscribe to a channel. "Unstable" is maybe overated, I meant that it was something causing more delays for that kind of worker.

So, to please everyone (including the people who were used to that behavior), it would be a good idea to add a `autoDisconnectPubSub` option [default=true] to let everyone enjoy the most suitable behavior.

**Comment by thetutlage** [comment]
Yeah I agree. It will be nice to have long living connection. Want to do a PR for same?

**Comment by richie3366** [comment]
Yep!

To recap & agree on the targeted PR:
- fix the check for existing (p)channel to only disconnect when there are no subscribed channel+pchannels 
- fix the disconnect process by nulling `this.subscriberConnection` after closing the connection
- add a `autoDisconnectPubSub`(*) option with a default value to `true`, in order to prevent the discussed behavior

Is it ok for you?

(*) if you think of a better name, tell me

**Comment by thetutlage** [comment]
The issue is fixed on newer `3.0.0` release, now it doesn't disconnect the connection at all. Also, the newer version is compatible with Adonisv5 (which isn't released yet)
